stingray/stats.py

```diff
@@ -0,0 +1,268 @@
+    p1 = stats.chi2.sf(stat, (nbin - 1))
+    return p_multitrial_from_single_trial(p1, ntrial)
+
+
+def fold_detection_level(nbin, epsilon=0.01, ntrial=1):
+    """Return the detection level for a folded profile.
+
+    See Leahy et al. (1983).
+
+    Parameters
+    ----------
+    nbin : int
+        The number of bins in the profile
+    epsilon : float, default 0.01
+        The fractional probability that the signal has been produced by noise
+
+    Other Parameters
+    ----------------
+    ntrial : int
+        The number of trials executed to find this profile
+
+    Returns
+    -------
+    detlev : float
+        The epoch folding statistics corresponding to a probability
+        epsilon * 100 % that the signal has been produced by noise
+    """
+    epsilon = p_single_trial_from_p_multitrial(epsilon, ntrial)
+    return stats.chi2.isf(epsilon.astype(np.double), nbin - 1)
+
+
+def z2_n_probability(z2, n=2, ntrial=1, n_summed_spectra=1):
+    """Calculate the probability of a certain folded profile, due to noise.
+
+    Parameters
+    ----------
+    z2 : float
+        A Z^2_n statistics value
+    n : int, default 2
+        The ``n`` in $Z^2_n$ (number of harmonics, including the fundamental)
+
+    Other Parameters
+    ----------------
+    ntrial : int
+        The number of trials executed to find this profile
+    n_summed_spectra : int
+        Number of Z_2^n periodograms that were averaged to obtain z2
+
+    Returns
+    -------
+    p : float
+        The probability that the Z^2_n value has been produced by noise
+    """
+    epsilon_1 = stats.chi2.sf(z2 * n_summed_spectra,
+                              2 * n * n_summed_spectra)
+    epsilon = p_multitrial_from_single_trial(epsilon_1, ntrial)
+
+    return epsilon
+
+
+def z2_n_detection_level(n=2, epsilon=0.01, ntrial=1, n_summed_spectra=1):
+    """Return the detection level for the Z^2_n statistics.
+
+    See Buccheri et al. (1983), Bendat and Piersol (1971).
+
+    Parameters
+    ----------
+    n : int, default 2
+        The ``n`` in $Z^2_n$ (number of harmonics, including the fundamental)
+    epsilon : float, default 0.01
+        The fractional probability that the signal has been produced by noise
+
+    Other Parameters
+    ----------------
+    ntrial : int
+        The number of trials executed to find this profile
+    n_summed_spectra : int
+        Number of Z_2^n periodograms that are being averaged
+
+    Returns
+    -------
+    detlev : float
+        The epoch folding statistics corresponding to a probability
+        epsilon * 100 % that the signal has been produced by noise
+    """
+
+    epsilon = p_single_trial_from_p_multitrial(epsilon, ntrial)
+    retlev = stats.chi2.isf(epsilon.astype(np.double),
+                            2 * n_summed_spectra * n) / (n_summed_spectra)
+
+    return retlev
+
+
+def pds_probability(level, ntrial=1, n_summed_spectra=1, n_rebin=1):
+    r"""Give the probability of a given power level in PDS.
```

Comment:
Okay, so, this function seems to be in direct conflict with `classical_pvalue`. Before merging this, I would like to understand which formalism we should be using, and make sure that we have one set of functions/methods that computes these statistics.

Reply:
Ok, now I understand better. The Groth implementation just boils down to `chi2` statistics in the absence of source variability. Since here we are interested in the noise powers, then the two formulations are equivalent!

To demonstrate this:
```
>>> from stingray.powerspectrum import classical_pvalue
>>> from stingray.stats import pds_probability
>>> import numpy as np
>>> powers = np.random.uniform(2, 30, 10)
>>> np.isclose([classical_pvalue(p, 1) for p in powers], pds_probability(powers))                                                                                                          
array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])
>>> np.isclose([classical_pvalue(p, 30) for p in powers], pds_probability(powers, n_summed_spectra=30))                                                                                    
array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])
```

But then:
1. `pds_probability` accepts arrays of power level out of the box
2. It has an `ntrial` option to account for the number of spectral bins, something classical_pvalue currently does not. In fact, the trial correction  performed in `Powerspectrum.classical_significances()` at the moment looks kind of weird.

Reply:
Note that the Bonferroni correction `(epsilon/ntrial)` is just an approximation. `p_single_trial_from_p_multitrial` uses `p = 1 - (1 - P)^(1/n)`, which is the exact correction for the number of trial using the binomial distribution.